substrate/pkg/controller/substrate/cluster/instance.go

```diff
@@ -20,68 +20,106 @@ import (
-		VPCZoneIdentifier: aws.String(strings.Join(substrate.Status.PublicSubnetIDs, ",")),
-		Tags: []*autoscaling.Tag{{
-			Key:               aws.String(discovery.OwnerTagKey),
-			Value:             aws.String(substrate.Name),
-			PropagateAtLaunch: aws.Bool(true),
-		}, {
-			Key:               aws.String("Name"),
-			Value:             discovery.Name(substrate),
-			PropagateAtLaunch: aws.Bool(true),
-		}},
-	}); err != nil {
-		if strings.Contains(err.Error(), "Invalid IAM Instance Profile name") { // Instance Profile can take several seconds to propagate
+	instancesOutput, err := i.EC2.DescribeInstancesWithContext(ctx, &ec2.DescribeInstancesInput{Filters: discovery.Filters(substrate)})
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("describing instances, %w", err)
+	}
+	for _, reservation := range instancesOutput.Reservations {
+		for _, instance := range reservation.Instances {
+			if aws.StringValue(instance.State.Name) == ec2.InstanceStateNameRunning || aws.StringValue(instance.State.Name) == ec2.InstanceStateNamePending {
+				for _, tag := range instance.Tags {
+					if aws.StringValue(tag.Key) == "aws:ec2launchtemplate:version" && aws.StringValue(tag.Value) == aws.StringValue(substrate.Status.Cluster.LaunchTemplateVersion) {
+						logging.FromContext(ctx).Infof("Found instance %s", aws.StringValue(instance.InstanceId))
+						return reconcile.Result{}, nil
+					}
+				}
+			}
+		}
+	}
+	overrides := []*ec2.FleetLaunchTemplateOverridesRequest{}
+	for _, subnet := range substrate.Status.PublicSubnetIDs {
+		overrides = append(overrides, &ec2.FleetLaunchTemplateOverridesRequest{SubnetId: aws.String(subnet)})
+	}
+	createFleetOutput, err := i.EC2.CreateFleetWithContext(ctx, &ec2.CreateFleetInput{
+		Type: aws.String(ec2.FleetTypeInstant),
+		LaunchTemplateConfigs: []*ec2.FleetLaunchTemplateConfigRequest{{
+			Overrides: overrides,
+			LaunchTemplateSpecification: &ec2.FleetLaunchTemplateSpecificationRequest{
+				LaunchTemplateName: discovery.Name(substrate),
+				Version:            substrate.Status.Cluster.LaunchTemplateVersion,
+			}},
+		},
+		TargetCapacitySpecification: &ec2.TargetCapacitySpecificationRequest{
+			DefaultTargetCapacityType: aws.String(ec2.DefaultTargetCapacityTypeOnDemand),
+			TotalTargetCapacity:       aws.Int64(1),
+		},
+		TagSpecifications: discovery.Tags(substrate, ec2.ResourceTypeInstance, discovery.Name(substrate)),
+		OnDemandOptions:   &ec2.OnDemandOptionsRequest{AllocationStrategy: aws.String(ec2.FleetOnDemandAllocationStrategyLowestPrice)},
+	})
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("creating fleet, %w", err)
+	}
+	for _, err := range createFleetOutput.Errors {
+		if strings.Contains(aws.StringValue(err.ErrorMessage), "Invalid IAM Instance Profile name") {
 			return reconcile.Result{Requeue: true}, nil
 		}
-		if err.(awserr.Error).Code() != autoscaling.ErrCodeAlreadyExistsFault {
-			return reconcile.Result{}, fmt.Errorf("creating autoscaling group, %w", err)
+	}
```

Comment:
There can be other errors also in `createFleetOutput.Errors`? Looks like we should be returning an error if any other error exist in `createFleetOutput.Errors`?